pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,463 @@
+package synthesize
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	rv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/storage/names"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+)
+
+var ObjectTypeNotFoundInBundleError = errors.New("object kind not found in bundle")
+
+type synthesize struct {
+	manifestDir           string
+	metadataDir           string
+	nonInferableCSV       *NonInferableCSV
+	unstructuredObjectMap map[string][]unstructured.Unstructured // map[kind]: *array of unstructured objects
+	existingCSVPath       string
+	synthesizedCSV        *operatorsv1alpha1.ClusterServiceVersion
+	apiDescriptionMap     map[registry.GroupVersionKind]*ApiDescription
+	logger                *log.Entry
```

Comment:
Nit: This could accept https://github.com/sirupsen/logrus/blob/v1.6.0/logrus.go#L139 instead (which is implemented by Entry and Log).